consensus/istanbul/backend/engine_test.go

```diff
@@ -506,3 +577,170 @@ func TestWriteCommittedSeals(t *testing.T) {
+	stakingInfo := &reward.StakingInfo{
+		BlockNum: blockNumber,
+	}
+	for idx, key := range keys {
+		addr := crypto.PubkeyToAddress(key.PublicKey)
+
+		pk, _ := crypto.GenerateKey()
+		rewardAddr := crypto.PubkeyToAddress(pk.PublicKey)
+
+		stakingInfo.CouncilNodeAddrs = append(stakingInfo.CouncilNodeAddrs, addr)
+		stakingInfo.CouncilStakingAmounts = append(stakingInfo.CouncilStakingAmounts, amounts[idx])
+		stakingInfo.CouncilRewardAddrs = append(stakingInfo.CouncilRewardAddrs, rewardAddr)
+	}
+	return stakingInfo
+}
+
+func checkInValidators(target common.Address, list []istanbul.Validator) bool {
+	for _, val := range list {
+		if target == val.Address() {
+			return true
+		}
+	}
+	return false
+}
+
+func isAllDemoted(minimumStaking uint64, stakingAmounts []uint64) bool {
+	for _, val := range stakingAmounts {
+		if val >= minimumStaking {
+			return false
+		}
+	}
+	return true
+}
+
+func TestSnapshot(t *testing.T) {
+	type testcase struct {
+		stakingAmounts        []uint64 // test staking amounts of each validator
+		isIstanbulCompatible  bool     // whether or not if the inserted block is istanbul compatible
+		expectedValidatorsNum int      // the number of expected validators
+		expectedDemotedNum    int      // the number of expected demoted validators
+	}
+
+	testcases := []testcase{
+		// The following testcases are the ones before istanbul incompatible change
+		{
+			[]uint64{5000000, 5000000, 5000000, 5000000},
+			false,
+			4,
+			0,
+		},
+		{
+			[]uint64{5000000, 5000000, 5000000, 6000000},
+			false,
+			4,
+			0,
+		},
+		{
+			[]uint64{5000000, 5000000, 6000000, 6000000},
+			false,
+			4,
+			0,
+		},
+		{
+			[]uint64{5000000, 6000000, 6000000, 6000000},
+			false,
+			4,
+			0,
+		},
+		{
+			[]uint64{6000000, 6000000, 6000000, 6000000},
+			false,
+			4,
+			0,
+		},
+		// The following testcases are the ones after istanbul incompatible change
+		{
+			[]uint64{5000000, 5000000, 5000000, 5000000},
+			true,
+			4,
+			0,
+		},
+		{
+			[]uint64{5000000, 5000000, 5000000, 6000000},
+			true,
+			1,
+			3,
+		},
+		{
+			[]uint64{5000000, 5000000, 6000000, 6000000},
+			true,
+			2,
+			2,
+		},
+		{
+			[]uint64{5000000, 6000000, 6000000, 6000000},
+			true,
+			3,
+			1,
+		},
+		{
+			[]uint64{6000000, 6000000, 6000000, 6000000},
+			true,
+			4,
+			0,
+		},
+		{
+			[]uint64{5500001, 5500000, 5499999, 0},
+			true,
+			2,
+			2,
+		},
+	}
+
+	testNum := 4
+	ms := uint64(5500000)
+	configItems := makeSnapshotTestConfigItems()
+	configItems = append(configItems, minimumStake(new(big.Int).SetUint64(ms)))
+	for _, tc := range testcases {
+		if tc.isIstanbulCompatible {
+			configItems = append(configItems, istanbulCompatibleBlock(new(big.Int).SetUint64(0)))
+		}
+		chain, engine := newBlockChain(testNum, configItems...)
+
+		stakingInfo := makeFakeStakingInfo(0, nodeKeys, tc.stakingAmounts)
+		reward.SetTestStakingManagerWithStakingInfoCache(stakingInfo)
```

Comment:
`SetTestStakingManagerWithStakingInfoCache` changes a global variable, stakingManager. I am afraid that the changed value affects other tests executed later. How about changing back to the original value when the test is finished.

Reply:
@aidan-kwon Reflected :) PTAL.